packages/serverpod/lib/src/database/concepts/exceptions.dart

```diff
@@ -0,0 +1,44 @@
+/// Exception thrown when an error occurs in the database.
+abstract interface class DatabaseException implements Exception {
+  /// Returns the message of the exception.
+  String get message;
+
+  @override
+  String toString() => 'DatabaseException: $message';
+}
+
+/// Exception thrown when an exception occurs during a database query.
+abstract interface class DatabaseQueryException implements DatabaseException {
+  /// Returns the error code of the exception.
+  String? get code;
+
+  /// Additional details if provided by the database.
+  String? get detail;
+
+  /// A hint on how to remedy an error, if provided by the database.
+  String? get hint;
+
+  /// Returns the name of the table where the error occurred.
+  String? get tableName;
+
+  /// Returns the name of the column where the error occurred.
+  String? get columnName;
+
+  /// Returns the name of the constraint that was violated.
+  String? get constraintName;
+
+  /// Returns the position in the query where the error occurred.
+  int? get position;
+}
+
+/// Exception thrown when an insert row operation fails.
+abstract interface class DatabaseInsertRowException
```

Comment:
Discussion: Should we really have different exceptions for different operation types? The exception typically represents the kind of error, not the kind of operation that the error prevented. 

For example, the duplicate value error can happen both on update and insert operations. No such key error can happen both on select, update, and delete operations.

I would propose to have exception subclasses for categories of errors - if we need them at all at this point. Typical categories could then be for example constraint violation (including duplicate and null-for-non-nullable field) and no-such-record (when refering to a specific record id in select, update, or delete).

Reply:
Good points here, I was thinking about this as well.

Reply:
IMO this principle is valid for our internal exceptions as well. The example you linked to perhaps shows need for having exception types for error categories rather than operation categories. In that example, there is a significant difference between the number of affected rows being zero, meaning "no such row / id found", and greater than 1, meaning ambiguous/wrong filter input. Both are presumably "caller issues" rather than DB insert issues?

Reply:
Discussed IRL, the principle of the argument is correct here. These exceptions should probably have been defined on a different level. Unfortunately, this would be a breaking change to modify, so we can't really fix it at this point.